cli/src/main/java/com/liferay/blade/cli/command/WatchCommand.java

```diff
@@ -75,102 +98,307 @@ private void _addError(String prefix, String msg) {
 
-		File baseDir = new File(baseArgs.getBase());
+		pathMatchers.add(pathMatcher);
+	}
 
-		WorkspaceProvider workspaceProvider = bladeCLI.getWorkspaceProvider(baseDir);
+	private String _getGradlePath(Path path, Path basePath) {
+		Path relativePath = basePath.relativize(path);
 
-		File workspaceDir = workspaceProvider.getWorkspaceDir(bladeCLI);
+		String gradlePath = ":" + relativePath.toString();
 
-		Set<String> projectPaths = projectOutputFiles.keySet();
+		gradlePath = gradlePath.replaceAll(File.separator, ":");
+
+		return gradlePath;
+	}
 
-		for (String projectPath : projectPaths) {
-			File projectDir = new File(workspaceDir, projectPath.replaceAll(":", File.separator));
+	private Path _getGradleProjectPath(Path basePath, Path path, Map<String, Path> projectPaths) {
+		String gradlePath = _getGradlePath(path, basePath);
 
-			File watchOutputFile = new File(projectDir, "build/installedBundleId");
+		for (Map.Entry<String, Path> entry : projectPaths.entrySet()) {
+			String projectPath = entry.getKey();
 
-			if (watchOutputFile.exists()) {
-				watchOutputFile.delete();
+			if (gradlePath.startsWith(projectPath)) {
+				return projectPaths.get(projectPath);
 			}
 		}
+
+		return basePath;
 	}
 
-	private void _uninstallBundles(Map<String, Set<File>> projectOutputFiles) {
-		BladeCLI bladeCLI = getBladeCLI();
+	private List<PathMatcher> _getPathMatchers(Path baseDirPath, String... patterns) {
+		FileSystem fileSystem = FileSystems.getDefault();
+
+		List<PathMatcher> pathMatchers = new ArrayList<>(patterns.length);
+
+		String patternPrefix = baseDirPath.toAbsolutePath() + File.separator;
+
+		if (File.separatorChar != '/') {
+			patternPrefix = patternPrefix.replace(File.separatorChar, '/');
+		}
 
-		BaseArgs baseArgs = bladeCLI.getArgs();
+		for (String pattern : patterns) {
+			if (pattern.startsWith("**/")) {
+				String absolutePattern = patternPrefix + pattern.substring(3);
 
-		File baseDir = new File(baseArgs.getBase());
+				_addPathMatcher(pathMatchers, fileSystem, absolutePattern);
+			}
 
-		WorkspaceProvider workspaceProvider = bladeCLI.getWorkspaceProvider(baseDir);
+			String absolutePattern = patternPrefix + pattern;
 
-		File workspaceDir = workspaceProvider.getWorkspaceDir(bladeCLI);
+			_addPathMatcher(pathMatchers, fileSystem, absolutePattern);
+		}
+
+		return pathMatchers;
+	}
 
-		Set<String> projectPaths = projectOutputFiles.keySet();
+	private Map<String, Path> _getProjectPaths(
+			final Path watchPath, List<String> projectPaths, List<String> ignorePaths)
+		throws Exception {
 
-		try (final GogoShellClient client = new GogoShellClient()) {
-			for (String projectPath : projectPaths) {
-				File projectDir = new File(workspaceDir, projectPath.replaceAll(":", File.separator));
+		final Map<String, Path> foundProjectPaths = new HashMap<>();
 
-				File watchOutputFile = new File(projectDir, "build/installedBundleId");
+		FileSystem fileSystem = FileSystems.getDefault();
```

Comment:
Anywhere using `FileSystem` should be wrapped in a _try-with-resources_ block to ensure the `FileSystem` object gets closed, as it implements `AutoCloseable`.

Reply:
Thanks for tip.